R/derive_advs_params.R

```diff
@@ -0,0 +1,600 @@
+#'
+#'
+#' @return The input dataset with the new parameter added. Note, a variable will only
+#'    be populated in the new parameter rows if it is specified in \code{by_vars}.
+#'
+#' @family internal
+#' @keywords internal
+derive_param_ratio <- function(dataset,
+                               by_vars,
+                               numerator_code,
+                               denominator_code,
+                               set_values_to,
+                               constant_numerator = FALSE,
+                               constant_denominator = FALSE,
+                               filter = NULL,
+                               constant_by_vars = NULL,
+                               get_unit_expr = NULL) {
+  assert_vars(by_vars)
+  assert_data_frame(dataset, required_vars = exprs(!!!by_vars, PARAMCD, AVAL))
+  assert_character_scalar(numerator_code)
+  assert_character_scalar(denominator_code)
+  assert_varval_list(set_values_to, required_elements = "PARAMCD")
+  assert_param_does_not_exist(dataset, set_values_to$PARAMCD)
+  assert_logical_scalar(constant_numerator)
+  assert_logical_scalar(constant_denominator)
+  filter <- assert_filter_cond(enexpr(filter), optional = TRUE)
+  assert_vars(constant_by_vars, optional = TRUE)
+  get_unit_expr <- assert_expr(enexpr(get_unit_expr), optional = TRUE)
+
+  ### Default formula with no units conversion applied ----
+
+  ratio_formula <- expr(
+    !!sym(paste0("AVAL.", numerator_code)) /
+      !!sym(paste0("AVAL.", denominator_code))
+  )
+
+  ### If `get_unit_expr` provided then check units and enable units conversion ----
+
+  if (!missing(get_unit_expr) && !is.null(get_unit_expr)) {
+    param_units <- list()
+
+    # Check if units are the same within each of two parameters
+
+    for (param in c(numerator_code, denominator_code)) {
+      units_found <- dataset %>%
+        mutate(`_unit` = !!get_unit_expr) %>%
+        filter(PARAMCD == param) %>%
+        pull(`_unit`) %>%
+        unique()
+
+      if (length(units_found) != 1L) {
+        cli_abort(
+          "Multiple units {.val {units_found}} found for {.val {param}}.
+          Please review and update the units."
+        )
+      }
+
+      param_units[[param]] <- units_found
+    }
+
+    # If the input parameters are measured in different units
+    # but are convertible from one to another (and this kind of conversion supported)
+    # then modify the formula in order to perform units conversion on the fly
+
+    if (param_units[[denominator_code]] != param_units[[numerator_code]]) {
+      # Find conversion factor for denominator
+      conv_factor <- get_conv_factor(
+        param_units[[denominator_code]],
+        param_units[[numerator_code]]
+      )
+
+      if (!is.na(conv_factor)) {
+        ratio_formula <- expr(
+          !!sym(paste0("AVAL.", numerator_code)) / (
+            !!sym(paste0("AVAL.", denominator_code)) * !!conv_factor
+          )
+        )
```

Comment:
Unit conversion must be optional because there might be other ratios where units of two input parameters are not convertible, e.g. when if I derived Height (cm) to Weight (kg), I would not expect any conversion factor between cm and kg. The result is just X cm/kg. Why should it fail in this case?

Reply:
Good point!

I however still mean that we need a way to fail if there are unsupported units in the `waisthip` and `waistheight` functions. This is now the case, when you have implemented the unit assertion function. Feel free to resolve this conversation.

Reply:
See my replies in other conversations.

`get_unit_expr` is no longer optional in the `waisthip` and `waistheight` and the units are asserted with the list of supported ones.

Reply:
Sorry, my formulation was confusing. I agree, this appears to be handled now.